build/src/lib.rs

```diff
@@ -40,50 +76,148 @@ impl Binary {
-    cache_dir: Option<P>,
-) -> Result<impl Read, Error> {
+) -> Result<Cursor<Vec<u8>>> {
     let filename = binary.filename(lang_code);
 
-    let mut cache_path: Option<PathBuf> = None;
-
-    if let Some(dir) = cache_dir {
-        cache_path = Some(dir.as_ref().join(version).join(lang_code).join(&filename));
-    }
-
-    // if the file can be read, the data is already cached
-    if let Some(path) = &cache_path {
-        if let Ok(bytes) = fs::read(path) {
-            return Ok(Cursor::new(bytes));
-        }
-    }
-
     // ... otherwise, request the data from the URL ...
     let bytes = reqwest::blocking::get(&format!(
         "https://github.com/bminixhofer/nlprule/releases/download/{}/{}.gz",
         version, filename
     ))?
-    .error_for_status()?
+    .error_for_status()
+    .map_err(|e| {
+        if let Some(404) = e.status().map(|x| x.as_u16()) {
+            Error::BinariesNotFound
+        } else {
+            e.into()
+        }
+    })?
     .bytes()?;
 
     let mut gz = GzDecoder::new(&bytes[..]);
     let mut buffer = Vec::new();
     gz.read_to_end(&mut buffer)?;
 
-    // ... and then cache the data at the provided file, if one was found
-    if let Some(path) = &cache_path {
-        fs::create_dir_all(path.parent().expect("path must have parent"))?;
-        fs::write(path, &buffer)?;
+    Ok(Cursor::new(buffer))
+}
+
+fn construct_cache_path(
+    version: &str,
+    lang_code: &str,
+    binary: Binary,
+    cache_dir: Option<&PathBuf>,
+    transform_path_fn: Option<&Box<dyn TransformPathFn>>,
+) -> Result<Option<PathBuf>> {
+    let filename = binary.filename(lang_code);
+
+    cache_dir
+        .map(move |dir| {
+            let path = dir.join(version).join(lang_code).join(&filename);
+            Ok(if let Some(transform_path_fn) = transform_path_fn {
+                transform_path_fn(path)?
+            } else {
+                path
+            })
+        })
+        .transpose()
+}
+
+/// Returns a `dyn Read` type, that is either directly feed
+/// from an in-memory slice or from the on disk cache.
+/// If the on-disk cache is disabled or is not present,
+/// it will attempt to download it via [`obtain_binary_from_github_release`].
+/// The data is then written into `dest`.
+/// Also updates the cache.
+fn obtain_binary_cache_or_github(
+    version: &str,
+    lang_code: &str,
+    binary: Binary,
+    cache_dir: Option<&PathBuf>,
+    transform_path_fn: Option<&Box<dyn TransformPathFn>>,
+    transform_data_fn: Option<&Box<dyn TransformDataFn<Cursor<Vec<u8>>>>>,
+) -> Result<Cursor<Vec<u8>>> {
+    let cache_path = construct_cache_path(
+        version,
+        lang_code,
+        binary,
+        cache_dir.clone(),
+        transform_path_fn,
+    )?;
+
+    // if the file can be read, the data is already cached and the transform was applied before
+    if let Some(ref cache_path) = cache_path {
+        if let Ok(bytes) = fs::read(cache_path) {
+            if bytes.len() > 256 {
```

Comment:
Why 256?

Reply:
The same thing as with the other magic number, if we keep this please add a comment. I'm also fine with removing it.